### DISCUSSION TITLE
webots simulation help..

### DISCUSSION THREAD

**Discussion by thdshtm** [community/community#55451] [Programming Help]
### Body

Can I write a code in this python code that doesn't press the keyboard to make it move, but moves to coordinates after a certain period of time, or adds random coordinates to the last coordinate I entered so that it arrives at the last coordinate?

from controller import Robot, Camera, InertialUnit, GPS, Compass, Gyro, Motor, Keyboard, Emitter, Accelerometer
import math
import numpy as np
import struct
import csv
import random
import time
import os
import threading
import mavic2proHelper
from simple_pid import PID
# create the Robot instance.

robot = Robot()

class MyAccelerometer(Accelerometer):
    def getNoise(self, n):
        noise = []
        for i in range(n):
            noise.append(random.uniform(-1, 1))
        return noise

    def getAcceleration(self):
        noise = self.getNoise(3)
        noise_accel_x = self.ax + noise[0]
        noise_accel_y = self.ay + noise[1]
        noise_accel_z = self.az + noise[2]
        return (noise_accel_x, noise_accel_y, noise_accel_z)
        
class MyGyro(Gyro):
    def getNoise(self, n):
        noise = []
        for i in range(n):
            noise.append(random.uniform(-1, 1))
        return noise

    def getGyro(self):
        noise = self.getNoise(3)
        noise_gyro_x = self.ax + noise[0]
        noise_gyro_y = self.ay + noise[1]
        noise_gyro_z = self.az + noise[2]
        return (noise_gyro_x, noise_gyro_y, noise_gyro_z)

gyro = robot.getDevice('gyro')
accelerometer = robot.getDevice('accelerometer')
# get the time step of the current world.
timestep = int(robot.getBasicTimeStep())
keyboard = Keyboard()
keyboard.enable(timestep)
emitter = robot.getEmitter("emitter")
# send data
message = struct.pack('5f', 0.0,0.0,0.0, 0.0, 0.0)
emitter.send(message)
imu = robot.getInertialUnit("inertial unit")
imu.enable(timestep)
camera = robot.getCamera("camera")
camera.enable(timestep)
gps = robot.getGPS("gps")
gps.enable(timestep)
compass = robot.getCompass("compass")
compass.enable(timestep)
gyro = robot.getGyro("gyro")
gyro.enable(timestep)
accelerometer = robot.getAccelerometer("accelerometer")
accelerometer.enable(timestep)
camera_roll_motor = robot.getCamera("camera roll")
camera_pitch_motor = robot.getCamera("camera pitch")
front_left_motor = robot.getMotor("front left propeller")
front_right_motor = robot.getMotor("front right propeller")
rear_left_motor = robot.getMotor("rear left propeller")
rear_right_motor = robot.getMotor("rear right propeller")
motors = [front_left_motor, front_right_motor, rear_left_motor, rear_right_motor]
# arming
for i in range(4):
    motors[i].setPosition(float("inf"))
    motors[i].setVelocity(1.0)

k_vertical_thrust = 68.5
k_vertical_offset = 0.6
k_vertical_p = 3.0  
k_roll_p = 50.0
k_pitch_p = 30.0
target_altitude = 1.0 

roll_disturbance = 0
pitch_disturbance = 0
yaw_disturbance = 0
# target
target_z = 1
target_x = 0
target_y = 0
target_yaw = 0
i = 0

end_x, end_y, end_z = 3, 5, 7

current_x, current_y, current_z = target_x, target_y, target_z

def convert_to_pitch_roll(ex, ey, yaw):
    c, s = np.cos(yaw), np.sin(yaw)
    R = np.array(((c, -s), (s, c)))
    exy_ = np.matmul([ex, ey], R)
    return exy_[0], exy_[1]

if os.path.exists('D:\중국 드론\data.csv'):
    os.remove('D:\중국 드론\data.csv')

data = []
header = ['gyro_x', 'gyro_y', 'gyro_z', 'accel_x', 'accel_y', 'accel_z','noise_a_x','noise_a_y','noise_a_z']#헤더쓰기

with open('D:\중국 드론\data.csv', mode='a', newline='') as f:
    writer = csv.writer(f)
    writer.writerow(header) 

last_time = time.time()#=>현재 시간
                   
while robot.step(timestep) != -1:
    
    roll = imu.getRollPitchYaw()[0] + math.pi / 2.0
    pitch = imu.getRollPitchYaw()[1]
    yaw = imu.getRollPitchYaw()[2]
    altitude = gps.getValues()[1]
    px = gps.getValues()[0]
    py = gps.getValues()[2]
    roll_disturbance = 0.0
    pitch_disturbance = 0.0
    yaw_disturbance = 0.0
    roll_acceleration = gyro.getValues()[0]
    pitch_acceleration = gyro.getValues()[1]
    compassValues = compass.getValues()
    gyroValues = gyro.getValues()
    x = compassValues[0] + gyroValues[0]
    y = compassValues[1] + gyroValues[1]
    z = compassValues[2] + gyroValues[2]   
    accel_x =   accelerometer.getValues()[0]  
    accel_y =   accelerometer.getValues()[1]  
    accel_z =   accelerometer.getValues()[2]-9.81
    noise_accel_x =  accelerometer.getValues()[0] + accelerometer.getValues()[0]*random.uniform(-0.1,0.1) 
    noise_accel_y =  accelerometer.getValues()[1] + accelerometer.getValues()[1]*random.uniform(-0.1,0.1) 
    noise_accel_z =  accelerometer.getValues()[2] + accelerometer.getValues()[2]*random.uniform(-0.1,0.1)-9.81 
    
    #data추가하기
    if time.time() - last_time >= 0.05: # 0.5초 지나면 행 추가
        last_time = time.time() # last_time 갱신(?)
        row = [gyro.getValues()[0], gyro.getValues()[1], gyro.getValues()[2],
               accel_x, accel_y, accel_z,noise_accel_x,noise_accel_y,noise_accel_z]
        data.append(row)
        
           
    if len(data)>5:#리스트가 5개 이상이면 저장 
        with open('D:\중국 드론\data.csv', mode='a', newline='') as f:
            writer = csv.writer(f)
            writer.writerows(data)
        data = []
            
        
   # keyboard input
    key=keyboard.getKey()
    while key > 0:
        if key == Keyboard.UP:
            target_z += 0.1 + random.random() * 0.9
        elif key == Keyboard.DOWN:
            target_z -= 0.1 + random.random() * 0.9
        elif key == Keyboard.RIGHT:
            yaw_disturbance = 1.0
        elif key == Keyboard.LEFT:
            yaw_disturbance = -1.0
        elif key == ord('W'):
            target_x += 1.5
            target_y += 1.5
            target_z += 1.0
            time.sleep(0.1)
            print("X : %f, Y : %f, Z : %f" % (target_x, target_y, target_z))
        elif key == ord('S'):
            target_x += 0.1 + random.random() * 0.9
        elif key == ord('A'):
            target_y += 0.1 + random.random() * 0.9
        elif key == ord('D'):
            target_y -= 0.1 + random.random() * 0.9
        key = 0
        
   
    # Compute the roll, pitch, yaw and vertical inputs.
    pitch_err, roll_err = convert_to_pitch_roll(px - target_x, target_y - py, yaw)
    
    roll_input = k_roll_p * np.clip(roll, -1.0, 1.0) + roll_acceleration - roll_err
    pitch_input = k_pitch_p * np.clip(pitch, -1.0, 1.0) - pitch_acceleration - pitch_err
    # roll_input = k_roll_p * np.clip(roll, -1.0, 1.0) + roll_acceleration + roll_disturbance
    # pitch_input = k_pitch_p * np.clip(pitch, -1.0, 1.0) - pitch_acceleration + pitch_disturbance
    yaw_input = yaw_disturbance
    clamped_difference_altitude = np.clip(target_z - altitude + k_vertical_offset, -1.0, 1.0)
    vertical_input = k_vertical_p * math.pow(clamped_difference_altitude, 3.0)
    # Actuate the motors taking into consideration all the computed inputs.
    front_left_motor_input = k_vertical_thrust + vertical_input - roll_input - pitch_input + yaw_input
    front_right_motor_input = k_vertical_thrust + vertical_input + roll_input - pitch_input - yaw_input
    rear_left_motor_input = k_vertical_thrust + vertical_input - roll_input + pitch_input - yaw_input
    rear_right_motor_input = k_vertical_thrust + vertical_input + roll_input + pitch_input + yaw_input
    
    front_left_motor.setVelocity(front_left_motor_input)
    front_right_motor.setVelocity(-front_right_motor_input)
    rear_left_motor.setVelocity(-rear_left_motor_input)
    rear_right_motor.setVelocity(rear_right_motor_input)

    # send data
    message = struct.pack('5f', target_x, target_y, target_z, yaw, yaw_disturbance)
    emitter.send(message)
     #Enter here functions to send actuator commands, like:
     #motor.setPosition(10.0)
    pass

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by RamtinKosari** [comment]
Hi dear @thdshtm, Yes, you can modify the code to make the robot move to coordinates after a certain period of time or add random coordinates to the last entered coordinate. Here's an example of how you can achieve that:

First, remove the keyboard input section from the code since you won't be using keyboard input to control the robot's movement.

Next, define the desired target coordinates and a time interval after which the robot should move to the target or add random coordinates. For example:
```python
import time

target_interval = 5  # Time interval in seconds
target_x = 0
target_y = 0
target_z = 0

last_target_time = time.time()

while robot.step(timestep) != -1:
    current_time = time.time()
    if current_time - last_target_time >= target_interval:
        # Move to target or add random coordinates
        last_target_time = current_time
        # Generate random coordinates
        target_x += random.uniform(-1, 1)
        target_y += random.uniform(-1, 1)
        target_z += random.uniform(-1, 1)
        # Move to the target coordinates
        # Set the target coordinates for the robot to move towards
        # ...

    # Rest of the code to control the robot's movement
    # ...
```

**Reply by RamtinKosari** [reply]
Hope this solution helps you my friend
